pkg/status/conditions.go

```diff
@@ -0,0 +1,200 @@
+// components care about rather than requiring those properties to be inferred
+// from other observations. Once defined, the meaning of a Condition can not be
+// changed arbitrarily - it becomes part of the API, and has the same
+// backwards- and forwards-compatibility concerns of any other part of the API.
+type Condition struct {
+	Type               ConditionType          `json:"type"`
+	Status             corev1.ConditionStatus `json:"status"`
+	Reason             ConditionReason        `json:"reason,omitempty"`
+	Message            string                 `json:"message,omitempty"`
+	LastTransitionTime metav1.Time            `json:"lastTransitionTime,omitempty"`
+}
+
+// IsTrue Condition whether the condition status is "True".
+func (c Condition) IsTrue() bool {
+	return c.Status == corev1.ConditionTrue
+}
+
+// IsFalse returns whether the condition status is "False".
+func (c Condition) IsFalse() bool {
+	return c.Status == corev1.ConditionFalse
+}
+
+// IsUnknown returns whether the condition status is "Unknown".
+func (c Condition) IsUnknown() bool {
+	return c.Status == corev1.ConditionUnknown
+}
+
+// DeepCopy returns a deep copy of the condition
+func (c *Condition) DeepCopy() *Condition {
+	if c == nil {
+		return nil
+	}
+	out := *c
+	return &out
+}
+
+// Conditions is a set of Condition instances.
+//
+// +kubebuilder:validation:Type=array
+type Conditions map[ConditionType]Condition
+
+// NewConditions initializes a set of conditions with the given list of
+// conditions.
+func NewConditions(conds ...Condition) Conditions {
+	conditions := Conditions{}
+	for _, c := range conds {
+		conditions.SetCondition(c)
+	}
+	return conditions
+}
+
+// IsTrue searches the set of conditions for a condition with the given
+// ConditionType. If found, it returns `condition.IsTrue()`. If not found,
+// it returns false.
+func (conditions Conditions) IsTrue(t ConditionType) bool {
+	if condition, ok := conditions[t]; ok {
+		return condition.IsTrue()
+	}
+	return false
+}
+
+// IsFalse searches the set of conditions for a condition with the given
+// ConditionType. If found, it returns `condition.IsFalse()`. If not found,
+// it returns false.
+func (conditions Conditions) IsFalse(t ConditionType) bool {
+	if condition, ok := conditions[t]; ok {
+		return condition.IsFalse()
+	}
+	return false
+}
+
+// IsUnknown searches the set of conditions for a condition with the given
+// ConditionType. If found, it returns `condition.IsUnknown()`. If not found,
+// it returns true.
+func (conditions Conditions) IsUnknown(t ConditionType) bool {
+	if condition, ok := conditions[t]; ok {
+		return condition.IsUnknown()
+	}
+	return true
+}
+
+// SetCondition adds (or updates) the set of conditions with the given
+// condition. It returns a boolean value indicating whether the set condition
+// is new or was a change to the existing condition with the same type.
+func (conditions *Conditions) SetCondition(newCond Condition) bool {
```

Comment:
Do these methods need to receive a pointer to `Conditions`? It’s a map type which will be updated in the caller’s scope regardless of pointer or not.

Reply:
Ideally, we'd leave the pointer receiver. Consider this:

```go
var conditions status.Conditions
conditions.SetCondition(myCondition)
```

This code would panic because the underlying map hasn't actually been initialized. Without the pointer receiver, a caller would need to do this instead:

```go
var conditions status.Conditions = make(map[status.ConditionType]status.Condition)
conditions.SetCondition(myCondition)
```

So TL;DR, the reason for the pointer receiver is so that we can initialize the underlying map for the caller if it is nil.